src/tls/openssl/tls.c

```diff
@@ -1779,3 +1804,168 @@ SSL_CTX *tls_ssl_ctx(const struct tls *tls)
 
 	return tls->ctx;
 }
+
+
+static void tls_cert_destructor(void *arg)
+{
+	struct tls_cert *uc = arg;
+
+	mem_deref(uc->host);
+	X509_free(uc->x509);
+	EVP_PKEY_free(uc->pkey);
+	sk_X509_pop_free(uc->chain, X509_free);
+}
+
+
+/**
+ * Adds a certificate for Server Name Indication (SNI) based certificate
+ * selection. An incoming client hello may contain an SNI extension which
+ * is used to select a local server certificate
+ *
+ * @param tls   TLS context
+ * @param certf Filename of the certificate
+ * @param host  Hostname that should match the SNI from client hello
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int tls_add_certf(struct tls *tls, const char *certf, const char *host)
+{
+	struct tls_cert *uc;
+	BIO *bio = NULL;
+	int err = 0;
+
+	if (!tls || !certf)
+		return EINVAL;
+
+	uc = mem_zalloc(sizeof(*uc), tls_cert_destructor);
+	if (!uc)
+		return ENOMEM;
+
+	if (str_isset(host)) {
+		err = str_dup(&uc->host, host);
+		if (err)
+			goto out;
+	}
+
+	bio = BIO_new_file(certf, "r");
+	if (!bio) {
+		err = EIO;
+		goto out;
+	}
+
+	uc->x509 = PEM_read_bio_X509(bio, NULL, NULL, NULL);
+	if (!uc->x509) {
+		DEBUG_WARNING("Can't read certificate from file: %s\n", certf);
+		err = ENOTSUP;
+		goto out;
+	}
+
+	while (1) {
+		X509 *ca = PEM_read_bio_X509(bio, NULL, 0, NULL);
+		if (!ca)
+			break;
+
+		if (!uc->chain)
+			uc->chain = sk_X509_new_null();
+
+		if (!uc->chain) {
+			err = ENOMEM;
+			goto out;
+		}
+
+		if (!sk_X509_push(uc->chain, ca)) {
+			err = ENOMEM;
+			goto out;
+		}
+	}
+
+	BIO_reset(bio);
```

Comment:
is this line needed ?

Reply:
Yes. See https://github.com/baresip/re/pull/596#discussion_r1046782443!

These steps are done in `tls_add_certf()`:
- first read certificate
- read potentially following CAs
- reset the BIO and read the key anywhere in the file